src/pipeline.py

```diff
@@ -52,19 +53,22 @@ def run(self):
             print(f"Received work {work}")
             executor = work.executor
             item = work.item
-            while executor is not None:
-                try:
-                    if executor.is_stateful():
-                        with executor.get_lock():
+            try:
+                while executor is not None:
+                    try:
+                        if executor.is_stateful():
+                            with executor.get_lock():
+                                item = executor.run(item)
+                        else:
                             item = executor.run(item)
-                    else:
-                        item = executor.run(item)
-                except RuntimeError as e:
-                    print(e)
-                    break
-                executor = executor.get_next()
-            print("Done processing work")
-
+                    except RuntimeError as e:
+                        print(e)
+                        break
+                    executor = executor.get_next()
+                print("Done processing work")
+            except CancelSignal:
+                item.is_cancelled = True
+                database_access.update_metadata(item)
```

Comment:
Can you add a check for the executor setting is_cancelled to True, but did not throw an CancelSignal?

Reply:
Why would we need this check? If you want to cancel an item, you should always throw the signal. If the item was already cancelled, it would not be downloaded